File: Project_Golang_String/mysolution/mysolution.go
package mysolution

import (
	"fmt"
	"io/ioutil"
	"os"
	"regexp"
	"strconv"
	"strings"
	"time"

	"./searchcleartextString"
	"./timeformatter"
)

var startTime int

// ClearTextMaxLength ... 平文の最大文字数
var ClearTextMaxLength int

// Start ...
func Start(openFileName string, threadCount int, searchMaxLength int, searchMode int, enableMuiltiThread bool, enableDebug bool) {
	// ハッシュ文字列が保存されたファイルの読み込み
	fp, _ := os.Open(openFileName)
	readBytes, _ := ioutil.ReadAll(fp)
	readText := string(readBytes)

	// コメント部の削除
	rep := regexp.MustCompile("#.*\n")
	readText = rep.ReplaceAllString(readText, "")
	rep = regexp.MustCompile("//.*\n")
	readText = rep.ReplaceAllString(readText, "")

	// ハッシュアルゴリズムとハッシュ文字列の分離
	temp := strings.Trim(readText, "")
	flds := strings.Split(temp, ":")

	algorithm := flds[0]
	targetHashedText := strings.Replace(flds[1], " ", "", -1)
	targetHashedText = strings.Replace(targetHashedText, "\r", "", -1)
	targetHashedText = strings.Replace(targetHashedText, "\n", "", -1)

	// 検索する平文の最大文字列長
	ClearTextMaxLength = searchMaxLength

	t := time.Now()
	const layout = "2006/01/02 15:04:05"

	fmt.Println("=====================================================================================")
	fmt.Println("Date               : " + t.Format(layout))
	fmt.Println("algorithm          : " + algorithm)
	fmt.Println("target Hashed Text : " + targetHashedText)
	if enableMuiltiThread {
		fmt.Println("thread count       : " + strconv.Itoa(threadCount))
	} else {
		fmt.Println("multithread        : " + strconv.FormatBool(enableMuiltiThread))
	}
	fmt.Println("search max length  : " + strconv.Itoa(searchMaxLength))
	fmt.Println("=====================================================================================")

	timeStart := time.Now()

	// 総当たり検索実行
	search(targetHashedText, algorithm, threadCount, ClearTextMaxLength, searchMode, enableMuiltiThread, enableDebug)

	timeEnd := time.Now()
	fmt.Printf("Execute time: %.3f [ms]\n\n", timeEnd.Sub(timeStart).Seconds()*1000)
}

//-----------------------------------------------------------------------------//
// 元の文字列を検索
//-----------------------------------------------------------------------------//
func search(targetHashedText string, algorithm string, threadMax int, searchMaxLength int, searchMode int, enableMuiltiThread bool, enableDebug bool) {
	// 使用するスレッド数の指定チェック
	if threadMax != 1 &&
		threadMax != 2 &&
		threadMax != 4 &&
		threadMax != 8 &&
		threadMax != 16 {
		return
	}

	algorithmUpper := strings.Replace(algorithm, "-", "", -1)
	algorithmUpper = strings.ToUpper(algorithmUpper)

	var AlgorithmIndex int
	switch algorithmUpper {
	case "MD5":
		AlgorithmIndex = 0
	case "SHA1":
		AlgorithmIndex = 1
	case "SHA256":
		AlgorithmIndex = 2
	case "SHA386":
		AlgorithmIndex = 3
	case "SHA512":
		AlgorithmIndex = 4
	case "RIPED160":
		AlgorithmIndex = 5
	default:
		AlgorithmIndex = 2 // default .. "SHA256"
	}

	// １文字から指定した文字列長まで検索する。
	for targetStrLength := 1; targetStrLength <= searchMaxLength; targetStrLength++ {

		// 平文検索処理用インスタンスの生成
		searchcleartextString.InitSearchClearText(AlgorithmIndex, targetHashedText, targetStrLength, threadMax, 0, enableMuiltiThread, enableDebug)

		// 文字数iでの総当たり平文検索開始時刻を保存
		currentStartTime := time.Now()

		//---------------------------------------------------------------------//
		// 文字数iでの総当たり平文検索開始
		//---------------------------------------------------------------------//
		resultStr, resultStrLen := searchcleartextString.GetClearText(threadMax)

		// 総当たり平文検索終了時刻との差を取得
		//	ts := time.Now() - startTime
		ts := (time.Now()).Sub(currentStartTime).Seconds()

		//---------------------------------------------------------------------//
		// 文字数iでの総当たり平文検索終了
		//---------------------------------------------------------------------//
		if resultStrLen >= 0 {
			fmt.Println("元の文字列が見つかりました！\r\n" +
				"\r\n" +
				"結果 = " + resultStr + "\r\n" +
				"\r\n" +
				"解析時間 = " + timeformatter.TsFormat(ts))
			break
		} else {
			fmt.Println(timeformatter.TsFormat(ts) + " ... " + strconv.Itoa(targetStrLength) + "文字の組み合わせ照合終了")
		}
	}
}

File: Project_Python3_Bytes/mylibs/MySolution.py
# coding: utf-8

import datetime
import time
import re

""" import local liblarys """
from .SearchClearText import SearchClearText
from . import TimeFormatter

class MySolution:

    def __init__(self):
        self.startTime = None
        self.ClearTextMaxLength = 0


    def Main(self, open_FileName, thread_count, search_max_length, search_mode, use_multiThread, use_debug):
        # ハッシュ文字列が保存されたファイルの読み込み
        with open(open_FileName, mode='r') as f:
            lines = f.readlines()

        # コメント部の削除
        read_Text = ""
        for line in lines:
            line = re.sub('#.*\n', "", line)
            line = re.sub('//.*\n', "", line)
            read_Text += line

        # ハッシュアルゴリズムとハッシュ文字列の分離
        flds = read_Text.rstrip().split(':')
        algorithm = flds[0]
        target_hashed_text = flds[1].rstrip()

        # 検索する平文の最大文字列長
        ClearTextMaxLength = search_max_length

        if use_multiThread:
            print("=====================================================================================\n"
                "Date               : {0}\n"
                "algorithm          : {1}\n"
                "target Hashed Text : {2}\n"
                "Collation type     : []byte\n"
                "thread count       : {3}\n"
                "search max length  : {4}\n"
                "====================================================================================="
                .format("{0:%Y-%m-%d %H:%M:%S}".format(datetime.datetime.now()), algorithm, target_hashed_text, thread_count, search_max_length))
        else:
            print("=====================================================================================\n"
                "algorithm          : {0}\n"
                "target Hashed Text : {1}\n"
                "Use MultiThre      : {2}\n"
                "search max length  : {3}\n"
                "====================================================================================="
                .format(algorithm, target_hashed_text, use_multiThread, search_max_length))

        time0 = time.time()

        # 総当たり検索実行
        self.search(target_hashed_text, algorithm, thread_count, ClearTextMaxLength, search_mode, use_multiThread, use_debug)
        time1 = time.time()

        print("Execute time ... : {0:f} [s]\n".format(time1 - time0))
    

    def search(self, target_hashed_text, algorithm, threadMax, search_ClearText_MaxLength, search_mode, use_multiThread, use_debug):
        # 使用するスレッド数の指定チェック
        if threadMax != 1 \
        and threadMax != 2 \
        and threadMax != 4 \
        and threadMax != 8 \
        and threadMax != 16:
            return

        algorithm_upper = algorithm.replace("-", "").upper()
        if algorithm_upper =="MD5":
            Algorithm_Index = 0
        elif algorithm_upper == "SHA1":
            Algorithm_Index = 1
        elif algorithm_upper == "SHA256":
            Algorithm_Index = 2
        elif algorithm_upper == "SHA386":
            Algorithm_Index = 3
        elif algorithm_upper == "SHA512":
            Algorithm_Index = 4
        elif algorithm_upper == "RIPED160":
            Algorithm_Index = 5
        else:
            Algorithm_Index = 2  # default .. "SHA256"

        # 現在の時刻を取得
        startTime = time.time()

        # １文字から指定した文字列長まで検索する。
        for i in range(1, search_ClearText_MaxLength + 1):
            # 平文検索処理用インスタンスの生成
            searchClearText = SearchClearText(Algorithm_Index, target_hashed_text, i, threadMax, 0, use_multiThread, use_debug)

            # 文字数iでの総当たり平文検索開始時刻を保存
            # current_startTime = time.time()

            #---------------------------------------------------------------------#
            # 文字数iでの総当たり平文検索開始
            #---------------------------------------------------------------------#
            resultStr = searchClearText.Get_ClearText(threadMax)

            # 総当たり平文検索終了時刻との差を取得
            ts = time.time() - startTime

            #---------------------------------------------------------------------#
            # 文字数iでの総当たり平文検索終了
            #---------------------------------------------------------------------#
            if resultStr != None:
                print("元の文字列が見つかりました！\n"
                      "\n"
                      "結果 = {0}\n"
                      "\n"
                      "解析時間 = {1}".format(resultStr, TimeFormatter.format(ts)))
                break
            else:
                print("{0}  ... {1} 文字の組み合わせ照合終了".format(TimeFormatter.format(ts), i))

        # 平文検索処理用インスタンスを解放する。
        searchClearText = None

        if resultStr == None:
            # 見つからなかった場合
            print("見つかりませんでした。")
